nemo/collections/llm/quantization/quantizer.py

```diff
@@ -0,0 +1,332 @@
+        quant_cfg["quant_cfg"]["*output_quantizer"] = {
+            "num_bits": 8 if algorithm == "int8_sq" else (4, 3),
+            "axis": None,
+            "enable": enable_quant_kv_cache,
+        }
+        if algorithm == "int8_sq":
+            sq_alpha = self.quantization_config["sq_alpha"]
+            logging.info(f"Using int8_sq alpha = {sq_alpha}")
+            quant_cfg["algorithm"] = {"method": "smoothquant", "alpha": sq_alpha}
+
+        model = mtq.quantize(model, quant_cfg, forward_loop)
+
+        if decoder_type == "gptnext":
+            # We found squared_relu may have an under-calibration problem.
+            # Clamp the scaling_factor with a min threshold to avoid under-calibration.
+            match algorithm:
+                case "fp8":
+                    maxbound = 448
+                case "int8_sq":
+                    maxbound = 127
+                case _:
+                    maxbound = 0
+
+            model = mtq.postprocess_amax(
+                model, "*input_quantizer", lambda amax: torch.clamp(amax, min=0.01 * maxbound)
+            )
+
+        if dist.get_rank() == 0:
+            mtq.print_quant_summary(model)
+
+        return wrapped_model
+
+    def create_megatron_forward_loop(
+        self, get_dataloader, num_batches, seq_length=None, micro_batch_size=None, decoder_seq_length=None
+    ):
+        from megatron.core.pipeline_parallel.schedules import get_forward_backward_func
+
+        forward_backward_func = get_forward_backward_func()
+
+        def forward_step_func(data_iterator, model):
+            data = next(data_iterator)
+            batch_len, seq_len = data.shape
+            position_ids = torch.arange(seq_len, device=data.device).expand((batch_len, seq_len))
+            output_tensor = model(data, position_ids, None)
+
+            def _mock_loss_function(tensor):
+                return 0, {}
+
+            return output_tensor, _mock_loss_function
+
+        def loop(model):
+            dataloader = get_dataloader()
+            forward_backward_func(
+                forward_step_func=forward_step_func,
+                data_iterator=dataloader,
+                model=model,
+                num_microbatches=num_batches,
+                seq_length=seq_length,
+                micro_batch_size=micro_batch_size,
+                decoder_seq_length=decoder_seq_length,
+                forward_only=True,
+            )
+
+        return loop
+
+    def export(self, model: llm.GPTModel, nemo_checkpoint_path: Optional[str] = None) -> None:
+        assert self.export_config is not None, "Export config is not set"
+        # TODO: Add sample generate
+        # TODO: Support NeMo 2:
+        # if model.cfg.megatron_amp_O2:
+        #     model.model = unwrap_model(model.model, Float16Module)
+        export_dir = self.export_config["path"]
+        use_nfs_workspace = (model.trainer._fabric.__io__.num_nodes > 1) or (
+            model.config.pipeline_model_parallel_size > 1
```

Comment:
Does `model.config.pipeline_model_parallel_size > 1` also require using NFS? Previously it was not the case

Reply:
Discussed offline: there are issues with shared memory-based solution, so we need to use NFS currently